File: Backend_erp_cga/app/contexts/conformite/donnees_demo.py
"""Jeu de données de démonstration — § 13.5 du dossier de design.

Toutes ces données sont fictives. Elles sont employées **de manière cohérente d'un écran
à l'autre** : mêmes entreprises, mêmes montants, même période de référence — juillet
2026, la date du jour étant le 8 août 2026.

Ne jamais substituer de faux latin ni de valeurs aléatoires : les écrans doivent montrer
des chiffres qui s'additionnent.
"""

from __future__ import annotations

from datetime import date
from decimal import Decimal as D

from .modeles import (
    ContexteControle,
    Document,
    FactureAControler,
    LigneFacture,
    ModeReglement,
    Montants,
    Partie,
    RegimeEmetteur,
    Reglement,
    TypeDocument,
)

__all__ = ["FACTURES_DEMO", "facture_demo"]


def _adherent(denomination: str, niu: str, rccm: str) -> Partie:
    return Partie(
        denomination=denomination, niu=niu, niu_actif=True, rccm=rccm, regime=RegimeEmetteur.REEL
    )


#: F-2026-0412 — verdict attendu : MAJEUR (règlement en espèces au-delà du seuil).
_F0412 = FactureAControler(
    document=Document(
        type=TypeDocument.FACTURE_ACHAT,
        reference="F-2026-0412",
        date_emission=date(2026, 7, 12),
    ),
    emetteur=Partie(
        denomination="QUINCAILLERIE DU WOURI",
        niu="M053311224455R",
        niu_actif=True,
        rccm="RC/DLA/2015/B/0881",
        regime=RegimeEmetteur.REEL,
    ),
    destinataire=_adherent("SARL BATIMENT PLUS", "M081234567890P", "RC/DLA/2021/B/0977"),
    montants=Montants(total_ht=D(1_970_650), total_tva=D(379_350), total_ttc=D(2_350_000)),
    reglement=Reglement(mode=ModeReglement.ESPECES, date_reglement=date(2026, 7, 12)),
    lignes=[
        LigneFacture(designation="Ciment CPJ 42,5 — 250 sacs", montant_ht=D(1_250_000)),
        LigneFacture(designation="Fers à béton HA12 — 60 barres", montant_ht=D(720_650)),
    ],
)

#: F-2026-0413 — verdict attendu : CONFORME.
_F0413 = FactureAControler(
    document=Document(reference="F-2026-0413", date_emission=date(2026, 7, 15)),
    emetteur=Partie(
        denomination="TRANSPORT SAWA EXPRESS",
        niu="M042233445566T",
        niu_actif=True,
        rccm="RC/DLA/2016/B/0442",
        regime=RegimeEmetteur.REEL,
    ),
    destinataire=_adherent("SARL BATIMENT PLUS", "M081234567890P", "RC/DLA/2021/B/0977"),
    montants=Montants(total_ht=D(1_200_000), total_tva=D(231_000), total_ttc=D(1_431_000)),
    reglement=Reglement(mode=ModeReglement.VIREMENT, date_reglement=date(2026, 7, 18)),
    lignes=[
        LigneFacture(designation="Transport de matériaux Douala — Kribi", montant_ht=D(1_200_000))
    ],
)

#: F-2026-0414 — verdict attendu : BLOQUANT (NIU absent) + AVERTISSEMENT (doublon).
_F0414 = FactureAControler(
    document=Document(reference="F-2026-0414", date_emission=date(2026, 7, 18)),
    emetteur=Partie(
        denomination="NÉGOCE MOUNGO SARL",
        niu=None,
        niu_actif=None,
        regime=RegimeEmetteur.REEL,
    ),
    destinataire=_adherent("BOULANGERIE LA COLOMBE SARL", "M071122334455J", "RC/YAO/2020/B/0311"),
    montants=Montants(total_ht=D(450_000), total_tva=D(86_625), total_ttc=D(536_625)),
    reglement=Reglement(mode=ModeReglement.ORANGE_MONEY, date_reglement=date(2026, 7, 18)),
    lignes=[LigneFacture(designation="Farine de blé T55 — 90 sacs", montant_ht=D(450_000))],
    # Une facture du même fournisseur et du même montant a été reçue le 09/07/2026.
    contexte=ContexteControle(doublons_potentiels=1),
)

#: F-2026-0415 — verdict attendu : AVERTISSEMENT (désignation imprécise).
_F0415 = FactureAControler(
    document=Document(reference="F-2026-0415", date_emission=date(2026, 7, 22)),
    emetteur=Partie(
        denomination="IMPRIMERIE DU LITTORAL",
        niu="M061199887766D",
        niu_actif=True,
        rccm="RC/DLA/2014/B/0233",
        regime=RegimeEmetteur.REEL,
    ),
    destinataire=_adherent("AGRO-NKOLO SA", "M065544332211L", "RC/BAF/2018/B/0145"),
    montants=Montants(total_ht=D(3_400_000), total_tva=D(654_500), total_ttc=D(4_054_500)),
    reglement=Reglement(mode=ModeReglement.VIREMENT, date_reglement=date(2026, 7, 25)),
    lignes=[LigneFacture(designation="Travaux divers", montant_ht=D(3_400_000))],
)

#: F-2026-0416 — verdict attendu : CONFORME.
_F0416 = FactureAControler(
    document=Document(reference="F-2026-0416", date_emission=date(2026, 7, 28)),
    emetteur=Partie(
        denomination="SOCIÉTÉ DE GARDIENNAGE ALPHA SÉCURITÉ",
        niu="M077788990022S",
        niu_actif=True,
        rccm="RC/YAO/2013/B/0107",
        regime=RegimeEmetteur.REEL,
    ),
    destinataire=_adherent("CLINIQUE LE BON SAMARITAIN", "M093344556677N", "RC/YAO/2017/B/0892"),
    montants=Montants(total_ht=D(850_000), total_tva=D(163_625), total_ttc=D(1_013_625)),
    reglement=Reglement(mode=ModeReglement.CHEQUE, date_reglement=date(2026, 7, 30)),
    lignes=[
        LigneFacture(designation="Gardiennage — juillet 2026, 3 agents", montant_ht=D(850_000))
    ],
)

FACTURES_DEMO: dict[str, FactureAControler] = {
    facture.document.reference: facture
    for facture in (_F0412, _F0413, _F0414, _F0415, _F0416)
}


def facture_demo(reference: str) -> FactureAControler:
    if reference not in FACTURES_DEMO:
        connues = ", ".join(sorted(FACTURES_DEMO))
        raise KeyError(f"facture « {reference} » inconnue. Références disponibles : {connues}")
    return FACTURES_DEMO[reference]

File: Frontend_erp_cga/app/lib/donnees-demo.ts
/**
 * Jeu de données de démonstration — § 13 du dossier de design.
 *
 * ⚠️ TEMPORAIRE. Ces données alimentent les écrans dont le contexte backend
 * n'existe pas encore : B Portefeuille, C Collecte, F Obligations. Elles
 * disparaîtront écran par écran à mesure que ces contextes seront implémentés.
 * L'écran E02, lui, est déjà branché sur le vrai moteur de conformité — voir
 * `app/lib/api.ts`.
 *
 * Même période de référence partout : **juillet 2026**, la date du jour étant le
 * **9 août 2026**. Les chiffres doivent s'additionner d'un écran à l'autre :
 * jamais de valeur inventée au fil de l'eau.
 */

import type { Severite } from "../components/Gravite";
import type { Statut } from "../components/Montant";

export const AUJOURD_HUI = new Date("2026-08-09T00:00:00");
export const PERIODE_COURANTE = "juillet 2026";

export const CABINET = {
  nom: "Broad Range Consulting Group",
  agrement: "MINFI/DGI n° 00000048",
  agences: ["Douala Akwa", "Yaoundé Elig-Essono", "Bafoussam"],
};

/** § 13.2 — l'utilisateur connecté par défaut : chargée de clientèle. */
export const UTILISATEUR = {
  nom: "Aïcha MBALLA",
  initiales: "AM",
  role: "Chargée de clientèle",
  agence: "Douala Akwa",
};

export type Entreprise = {
  denomination: string;
  niu: string;
  rccm: string;
  regime: "REEL" | "IGS";
  ville: string;
  activite: string;
};

/** § 13.3 — les six adhérents de démonstration. */
export const ENTREPRISES: Entreprise[] = [
  {
    denomination: "SARL BATIMENT PLUS",
    niu: "M081234567890P",
    rccm: "RC/DLA/2021/B/0977",
    regime: "REEL",
    ville: "Douala Bonabéri",
    activite: "BTP",
  },
  {
    denomination: "ETS TCHOUMBA & FILS",
    niu: "P019876543210K",
    rccm: "RC/DLA/2019/A/1842",
    regime: "IGS",
    ville: "Douala Akwa",
    activite: "Commerce général",
  },
  {
    denomination: "BOULANGERIE LA COLOMBE SARL",
    niu: "M071122334455J",
    rccm: "RC/YAO/2020/B/0311",
    regime: "REEL",
    ville: "Yaoundé Elig-Essono",
    activite: "Agroalimentaire",
  },
  {
    denomination: "AGRO-NKOLO SA",
    niu: "M065544332211L",
    rccm: "RC/BAF/2018/B/0145",
    regime: "REEL",
    ville: "Bafoussam",
    activite: "Agro-industrie",
  },
  {
    denomination: "CABINET NGUEMA CONSEIL",
    niu: "P027788990011M",
    rccm: "RC/DLA/2022/A/2510",
    regime: "IGS",
    ville: "Douala Bonanjo",
    activite: "Profession libérale",
  },
  {
    denomination: "CLINIQUE LE BON SAMARITAIN",
    niu: "M093344556677N",
    rccm: "RC/YAO/2017/B/0892",
    regime: "REEL",
    ville: "Yaoundé Mvog-Ada",
    activite: "Santé",
  },
];

/** Dossiers ouverts récemment : le sélecteur les propose en tête. */
export const ENTREPRISES_RECENTES = [
  "SARL BATIMENT PLUS",
  "AGRO-NKOLO SA",
  "BOULANGERIE LA COLOMBE SARL",
];

/** Compteurs des pastilles de la barre latérale. */
export const COMPTEURS = {
  piecesEnAttente: 37,
  anomaliesBloquantes: 3,
};

/** § 8.1 — les quatre indicateurs du tableau de bord collaborateur. */
export type Indicateur = {
  libelle: string;
  valeur: string;
  detail: string;
  ton: "neutre" | "attention" | "alerte";
};

export const INDICATEURS: Indicateur[] = [
  { libelle: "Dossiers suivis", valeur: "18", detail: "sur 128 au cabinet", ton: "neutre" },
  {
    libelle: "Échéances sous 7 jours",
    valeur: "6",
    detail: "dont 5 au 15 août",
    ton: "attention",
  },
  {
    libelle: "Pièces en attente",
    valeur: "37",
    detail: "12 arrivées aujourd'hui",
    ton: "neutre",
  },
  {
    libelle: "Anomalies bloquantes",
    valeur: "3",
    detail: "non résolues · 1 depuis 22 jours",
    ton: "alerte",
  },
];

/** § 13.7 — échéances à venir. Le montant est estimé tant que rien n'est déposé. */
export type Echeance = {
  date: string;
  entreprise: string;
  obligation: string;
  montant: number;
  statut: Statut;
  piecesRecues: number;
  piecesAttendues: number;
  joursRestants: number;
};

export const ECHEANCES: Echeance[] = [
  {
    date: "2026-08-15",
    entreprise: "ETS TCHOUMBA & FILS",
    obligation: "Impôt Général Synthétique, 3e trimestre",
    montant: 275000,
    statut: "En retard",
    piecesRecues: 14,
    piecesAttendues: 19,
    joursRestants: -24,
  },
  {
    date: "2026-08-15",
    entreprise: "AGRO-NKOLO SA",
    obligation: "Déclaration de TVA, juillet 2026",
    montant: 12340000,
    statut: "À faire",
    piecesRecues: 31,
    piecesAttendues: 38,
    joursRestants: 6,
  },
  {
    date: "2026-08-15",
    entreprise: "SARL BATIMENT PLUS",
    obligation: "Déclaration de TVA, juillet 2026",
    montant: 4820000,
    statut: "En préparation",
    piecesRecues: 22,
    piecesAttendues: 24,
    joursRestants: 6,
  },
  {
    date: "2026-08-15",
    entreprise: "BOULANGERIE LA COLOMBE SARL",
    obligation: "Acompte d'impôt sur les sociétés, juillet",
    montant: 528000,
    statut: "Prête",
    piecesRecues: 17,
    piecesAttendues: 17,
    joursRestants: 6,
  },
  {
    date: "2026-08-15",
    entreprise: "CLINIQUE LE BON SAMARITAIN",
    obligation: "Cotisations CNPS, juillet 2026",
    montant: 1745000,
    statut: "Déclarée",
    piecesRecues: 12,
    piecesAttendues: 12,
    joursRestants: 6,
  },
  {
    date: "2026-09-30",
    entreprise: "AGRO-NKOLO SA",
    obligation: "Taxe sur la propriété foncière",
    montant: 890000,
    statut: "À faire",
    piecesRecues: 0,
    piecesAttendues: 2,
    joursRestants: 52,
  },
];

/** Anomalies récentes du portefeuille, ordonnées par gravité puis par enjeu. */
export type Anomalie = {
  entreprise: string;
  regle: string;
  code: string;
  severite: Severite;
  enjeu: number | null;
  /** Référence de la pièce : ouvre l'écran E02. */
  piece: string | null;
  anciennete: string;
};

export const ANOMALIES: Anomalie[] = [
  {
    entreprise: "BOULANGERIE LA COLOMBE SARL",
    regle: "NIU du fournisseur absent ou non reconnu",
    code: "FAC-ID-003",
    severite: "BLOQUANT",
    enjeu: 536625,
    piece: "F-2026-0414",
    anciennete: "22 j",
  },
  {
    entreprise: "AGRO-NKOLO SA",
    regle: "NIU du fournisseur absent ou non reconnu",
    code: "FAC-ID-003",
    severite: "BLOQUANT",
    enjeu: 812400,
    piece: null,
    anciennete: "18 j",
  },
  {
    entreprise: "AGRO-NKOLO SA",
    regle: "Facture au nom du dirigeant, non de l'entreprise",
    code: "FAC-ID-008",
    severite: "BLOQUANT",
    enjeu: 125775,
    piece: null,
    anciennete: "11 j",
  },
  {
    entreprise: "SARL BATIMENT PLUS",
    regle: "Règlement en espèces au-delà du seuil légal",
    code: "FAC-ACH-007",
    severite: "MAJEUR",
    enjeu: 379350,
    piece: "F-2026-0412",
    anciennete: "14 j",
  },
  {
    entreprise: "ETS TCHOUMBA & FILS",
    regle: "Règlement en espèces au-delà du seuil légal",
    code: "FAC-ACH-007",
    severite: "MAJEUR",
    enjeu: 117810,
    piece: null,
    anciennete: "10 j",
  },
  {
    entreprise: "AGRO-NKOLO SA",
    regle: "Écart de calcul entre lignes et total hors taxes",
    code: "FAC-CAL-002",
    severite: "MAJEUR",
    enjeu: 212400,
    piece: null,
    anciennete: "8 j",
  },
  {
    entreprise: "AGRO-NKOLO SA",
    regle: "Désignation de la prestation imprécise",
    code: "FAC-DOC-011",
    severite: "AVERTISSEMENT",
    enjeu: null,
    piece: "F-2026-0415",
    anciennete: "12 j",
  },
];

/** Dossiers incomplets dont l'échéance approche. */
export type DossierIncomplet = {
  entreprise: string;
  periode: string;
  manquantes: number;
  attendues: number;
  joursRestants: number;
  natureManquante: string;
};

export const DOSSIERS_INCOMPLETS: DossierIncomplet[] = [
  {
    entreprise: "AGRO-NKOLO SA",
    periode: "juillet 2026",
    manquantes: 7,
    attendues: 38,
    joursRestants: 6,
    natureManquante: "Relevés bancaires, 3 factures d'achat, bordereau CNPS",
  },
  {
    entreprise: "ETS TCHOUMBA & FILS",
    periode: "juillet 2026",
    manquantes: 5,
    attendues: 19,
    joursRestants: -24,
    natureManquante: "Relevé bancaire, factures de carburant",
  },
  {
    entreprise: "CABINET NGUEMA CONSEIL",
    periode: "juillet 2026",
    manquantes: 5,
    attendues: 14,
    joursRestants: 11,
    natureManquante: "Notes d'honoraires, justificatifs de déplacement",
  },
  {
    entreprise: "SARL BATIMENT PLUS",
    periode: "juillet 2026",
    manquantes: 2,
    attendues: 24,
    joursRestants: 6,
    natureManquante: "Facture rectificative NÉGOCE MOUNGO, justificatif de virement",
  },
];
